src/contract/contract.rs

```diff
@@ -185,39 +186,100 @@ impl<State: ExposedState> OutputAssignment<State> {
+#[derive(StrictType, StrictDumb, StrictEncode, StrictDecode)]
+#[strict_type(lib = LIB_NAME_RGB, dumb = { Self(1) })]
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_crate", rename_all = "camelCase")
+)]
+#[display(inner)]
+pub struct WitnessHeight(u32);
+
+impl WitnessHeight {
+    pub fn new(height: u32) -> Option<Self> {
+        match height {
+            0 => None,
+            height => Some(WitnessHeight(height)),
+        }
+    }
+
+    pub fn get(&self) -> NonZeroU32 { NonZeroU32::new(self.0).expect("invariant") }
+}
+
+/// RGB consensus information about the current mined height of a witness
+/// transaction defining the ordering of the contract state data.
+#[derive(Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Hash, Debug, Display)]
+#[derive(StrictType, StrictDumb, StrictEncode, StrictDecode)]
+#[strict_type(lib = LIB_NAME_RGB, tags = order)]
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_crate", rename_all = "camelCase")
+)]
+pub enum WitnessOrd {
+    #[display(inner)]
+    OnChain(WitnessHeight),
+
+    #[display("offchain")]
+    #[strict_type(dumb)]
+    OffChain,
+}
+
+impl WitnessOrd {
+    pub fn from_electrum_height(height: u32) -> Self {
+        WitnessHeight::new(height)
+            .map(WitnessOrd::OnChain)
+            .unwrap_or(OffChain)
+    }
+}
+
 /// Txid and height information ordered according to the RGB consensus rules.
 #[derive(Copy, Clone, Eq, PartialEq, Hash, Debug, Display)]
 #[derive(StrictType, StrictDumb, StrictEncode, StrictDecode)]
 #[strict_type(lib = LIB_NAME_RGB)]
-#[display("{height}/{txid}")]
+#[display("{ord}/{txid}")]
 #[cfg_attr(
     feature = "serde",
     derive(Serialize, Deserialize),
     serde(crate = "serde_crate", rename_all = "camelCase")
 )]
-pub struct OrderedTxid {
-    pub height: u32,
+pub struct WitnessAnchor {
+    pub ord: WitnessOrd,
     pub txid: Txid,
 }
 
-impl PartialOrd for OrderedTxid {
+impl PartialOrd for WitnessAnchor {
     fn partial_cmp(&self, other: &Self) -> Option<Ordering> { Some(self.cmp(other)) }
 }
 
-impl Ord for OrderedTxid {
+impl Ord for WitnessAnchor {
     fn cmp(&self, other: &Self) -> Ordering {
         if self == other {
             return Ordering::Equal;
         }
-        if self.height != other.height {
-            return self.height.cmp(&other.height);
+        if self.ord != other.ord {
+            return self.ord.cmp(&other.ord);
         }
         self.txid.cmp(&other.txid)
     }
 }
 
-impl OrderedTxid {
-    pub fn new(height: u32, txid: Txid) -> Self { OrderedTxid { height, txid } }
+impl WitnessAnchor {
+    pub fn new(ord: WitnessOrd, txid: Txid) -> Self { WitnessAnchor { ord, txid } }
+    pub fn with_mempool(txid: Txid) -> Self {
+        WitnessAnchor {
+            ord: WitnessOrd::OffChain,
+            txid,
+        }
+    }
+    pub fn with_electrum_height(height: u32, txid: Txid) -> Self {
```

Comment:
As we intend to support other explorers, like esplora, I think it's interesting to change this name.

```suggestion
    pub fn with_block_height(height: u32, txid: Txid) -> Self {
```

Reply:
See above